coderd/wsbuilder/wsbuilder.go

```diff
@@ -0,0 +1,741 @@
+}
+
+func (b Builder) State(state []byte) Builder {
+	// nolint: revive
+	b.state = stateTarget{explicit: &state}
+	return b
+}
+
+func (b Builder) Orphan() Builder {
+	// nolint: revive
+	b.state = stateTarget{orphan: true}
+	return b
+}
+
+func (b Builder) LogLevel(l string) Builder {
+	// nolint: revive
+	b.logLevel = l
+	return b
+}
+
+func (b Builder) Initiator(u uuid.UUID) Builder {
+	// nolint: revive
+	b.initiator = u
+	return b
+}
+
+func (b Builder) Reason(r database.BuildReason) Builder {
+	// nolint: revive
+	b.reason = r
+	return b
+}
+
+func (b Builder) LegacyParameterValues(p []codersdk.CreateParameterRequest) Builder {
+	// nolint: revive
+	b.legacyParameterValues = p
+	return b
+}
+
+func (b Builder) RichParameterValues(p []codersdk.WorkspaceBuildParameter) Builder {
+	// nolint: revive
+	b.richParameterValues = p
+	return b
+}
+
+// SetLastWorkspaceBuildInTx prepopulates the Builder's cache with the last workspace build.  This allows us
+// to avoid a repeated database query when the Builder's caller also needs the workspace build, e.g. auto-start &
+// auto-stop.
+//
+// CAUTION: only call this method from within a database transaction with RepeatableRead isolation.  This transaction
+// MUST be the database.Store you call Build() with.
+func (b Builder) SetLastWorkspaceBuildInTx(build *database.WorkspaceBuild) Builder {
+	// nolint: revive
+	b.lastBuild = build
+	return b
+}
+
+// SetLastWorkspaceBuildJobInTx prepopulates the Builder's cache with the last workspace build job.  This allows us
+// to avoid a repeated database query when the Builder's caller also needs the workspace build job, e.g. auto-start &
+// auto-stop.
+//
+// CAUTION: only call this method from within a database transaction with RepeatableRead isolation.  This transaction
+// MUST be the database.Store you call Build() with.
+func (b Builder) SetLastWorkspaceBuildJobInTx(job *database.ProvisionerJob) Builder {
+	// nolint: revive
+	b.lastBuildJob = job
+	return b
+}
+
+type BuildError struct {
+	// Status is a suitable HTTP status code
+	Status  int
+	Message string
+	Wrapped error
+}
+
+func (e BuildError) Error() string {
+	return e.Wrapped.Error()
+}
+
+func (e BuildError) Unwrap() error {
+	return e.Wrapped
+}
+
+// Build computes and inserts a new workspace build into the database.  If authFunc is provided, it also performs
+// authorization preflight checks.
+func (b *Builder) Build(
+	ctx context.Context,
+	store database.Store,
+	authFunc func(action rbac.Action, object rbac.Objecter) bool,
+) (
+	*database.WorkspaceBuild, *database.ProvisionerJob, error,
+) {
+	b.ctx = ctx
+
+	// Run the build in a transaction with RepeatableRead isolation, and retries.
+	// RepeatableRead isolation ensures that we get a consistent view of the database while
+	// computing the new build.  This simplifies the logic so that we do not need to worry if
+	// later reads are consistent with earlier ones.
+	var err error
+	for retries := 0; retries < 5; retries++ {
```

Comment:
Is it an improvement or did we have it in the current codebase?

Reply:
It's an improvement.  I've introduced RepeatableRead isolation to the transaction, which is a higher level of isolation than we previously used, but it does mean that transactions can fail if the rows you access get modified by something else before your transaction commits.

So, it introduces a new failure mode, and introduces retries to mitigate.

Reply:
Thanks for the explanation 👍 . I guess that we don't need any backoff policy here?